### PR TITLE
Patchwork Paper implementation using ROS 2.

### PR DESCRIPTION
**Pull request by akilpath**
Real-time LiDAR Ground Segmentation using Pathwork ([https://github.com/LimHyungTae/patchwork](url))

### CODE REVIEW

File: .github/templates/docker_context/docker_context.sh [hunk 1]
```diff
@@ -21,7 +21,7 @@ fi
 while read -r module; do
 
     # Retrieve docker compose service names
-    services=$(docker-compose -f "$module" config --services)
+    services=$(docker compose -f "$module" config --services)
```
Comment (Edwardius) [inline .github/templates/docker_context/docker_context.sh:24]: separate PR plz

File: docker/world_modeling/occupancy_segmentation/occupancy_segmentation.Dockerfile [hunk 2]
```diff
@@ -6,8 +6,8 @@ FROM ${BASE_IMAGE} as source
 WORKDIR ${AMENT_WS}/src
 
 # Copy in source code 
+# RUN git clone https://github.com/ros-perception/perception_pcl.git --branch 2.4.3
```
Comment (Edwardius) [inline docker/world_modeling/occupancy_segmentation/occupancy_segmentation.Dockerfile:9]: rm

File: docker/world_modeling/occupancy_segmentation/occupancy_segmentation.Dockerfile [hunk 3]
```diff
@@ -16,12 +16,14 @@ RUN apt-get -qq update && rosdep update && \
         | awk '{print $3}' \
         | sort  > /tmp/colcon_install_list
 
+RUN sudo apt-get install libeigen3-dev
+RUN sudo apt-get -y install libtbb-dev
```
Comment (Edwardius) [inline docker/world_modeling/occupancy_segmentation/occupancy_segmentation.Dockerfile:20]: place in dependencies stage

File: modules/dev_overrides/docker-compose.world_modeling.yaml [hunk 4]
```diff
@@ -13,7 +13,7 @@ services:
     image: "${WORLD_MODELING_HD_MAP_IMAGE}:build_${TAG}"
     command: tail -F anything
     volumes:
-      - ${MONO_DIR}/src/world_modeling:/home/ament_ws/src/hd_map
+      - ${MONO_DIR}/src/world_modeling:/home/bolty/ament_ws/src/hd_map
```
Comment (Edwardius) [inline modules/dev_overrides/docker-compose.world_modeling.yaml:16]: oh @VishGit1234 can this be a separate PR to fix all the volume directory mismatches?

File: src/world_modeling/occupancy_segmentation/include/occupancy_segmentation_core.hpp [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+#define MH -1.1
+
+#define MIN_NUM_POINTS 10
+#define NUM_SEED_POINTS 20
+#define TH_SEEDS 0.5
+#define UPRIGHTNESS_THRESH 0.5
+
+#define NUM_RINGS_OF_INTEREST 4
+#define SENSOR_HEIGHT 1.7
+#define GLOBAL_EL_THRESH 0
```
Comment (Edwardius) [inline src/world_modeling/occupancy_segmentation/include/occupancy_segmentation_core.hpp:28]: should be part of the Yaml config file, you can pass ROS loggers and param servers into core, that's ok

File: src/world_modeling/occupancy_segmentation/include/occupancy_segmentation_core.hpp [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+  // Size of buffer before processed messages are published.
+  static constexpr int BUFFER_CAPACITY = 10;
+
+  const int ZONE_RINGS[NUM_ZONES] = {2, 4, 4, 4};
+  const int ZONE_SECTORS[NUM_ZONES] = {16, 32, 54, 32};
+  const float FLATNESS_THR[NUM_ZONES] = {0.0005, 0.000725, 0.001, 0.001};
+  const float ELEVATION_THR[NUM_ZONES] = {0.523, 0.746, 0.879, 1.125};
+  const double lmins[NUM_ZONES] = {L_MIN, (7 * L_MIN + L_MAX) / 8, (3 * L_MIN + L_MAX) / 4,
+                                   (L_MIN + L_MAX) / 2};
+  const double lmaxs[NUM_ZONES] = {lmins[1], lmins[2], lmins[3], L_MAX};
```
Comment (Edwardius) [inline src/world_modeling/occupancy_segmentation/include/occupancy_segmentation_core.hpp:72]: seems like a bunch of magic numbers? correct me if these are rarely changed

File: src/world_modeling/occupancy_segmentation/include/occupancy_segmentation_core.hpp
Comment (Edwardius) [inline src/world_modeling/occupancy_segmentation/include/occupancy_segmentation_core.hpp:1]: generally, place class definitions and templates in the hpp file, but logic inside the cpp file.

I should be able to read the hpp file to get a good general idea of what the code is doing, and if I am more curious, I can take a look in the cpp file to see what the logic actually is.

File: src/world_modeling/occupancy_segmentation/src/occupancy_segmentation_core.cpp
Comment (Edwardius) [inline src/world_modeling/occupancy_segmentation/src/occupancy_segmentation_core.cpp:1]: very empty

File: src/world_modeling/occupancy_segmentation/include/occupancy_segmentation_node.hpp [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+  // Object that handles data processing and validation.
+  OccupancySegmentationCore<PointXYZIRT> _patchwork;
+  rclcpp::Subscription<sensor_msgs::msg::PointCloud2>::SharedPtr _subscriber;
+  rclcpp::Publisher<sensor_msgs::msg::PointCloud2>::SharedPtr _nonground_publisher;
+  rclcpp::Publisher<sensor_msgs::msg::PointCloud2>::SharedPtr _ground_publisher;
+
+  void subscription_callback(const sensor_msgs::msg::PointCloud2::SharedPtr lidar_cloud);
+};
+
+#endif  // TRANSFORMER_NODE_HPP_
```
Comment (Edwardius) [inline src/world_modeling/occupancy_segmentation/include/occupancy_segmentation_node.hpp:61]:
```suggestion
#endif  // OCCUPANCY_SEGMENTATION_NODE_HPP_
```

File: src/world_modeling/occupancy_segmentation/config/params.yaml
Comment (Edwardius) [inline src/world_modeling/occupancy_segmentation/config/params.yaml:1]: add all your constants for patchworks here

File: src/world_modeling/occupancy_segmentation/package.xml [hunk 8]
```diff
@@ -1,15 +1,21 @@
...
 <package format="3">
   <name>occupancy_segmentation</name>
   <version>0.0.0</version>
-  <description>TODO: Package description</description>
-  <maintainer email="e23zhou@uwaterloo.ca">bolty</maintainer>
-  <license>TODO: License declaration</license>
+  <description>A sample ROS package for pubsub communication</description>
 
+  <maintainer email="cmahalingam@watonomous.ca">Cheran Mahalingam</maintainer>
+  <license>TODO</license>
```
Comment (Edwardius) [inline src/world_modeling/occupancy_segmentation/package.xml:8]: make sure to make urself the maintainer

File: src/world_modeling/occupancy_segmentation/src/occupancy_segmentation_node.cpp [hunk 9]
```diff
@@ -0,0 +1,56 @@
+#include "occupancy_segmentation_node.hpp"
+#include <iostream>
+#include <memory>
+
+OccupancySegmentationNode::OccupancySegmentationNode() : Node("occupancy_segmentation") {
+  _subscriber = this->create_subscription<sensor_msgs::msg::PointCloud2>(
+      "/velodyne_points", 10,
```
Comment (Edwardius) [inline src/world_modeling/occupancy_segmentation/src/occupancy_segmentation_node.cpp:7]: specifying the topic name should be from the params.yaml file

File: src/world_modeling/occupancy_segmentation/src/occupancy_segmentation_node.cpp [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+
+  _ground_publisher = this->create_publisher<sensor_msgs::msg::PointCloud2>("/ground_points", 10);
+  _nonground_publisher =
+      this->create_publisher<sensor_msgs::msg::PointCloud2>("/nonground_points", 10);
+}
+
+void OccupancySegmentationNode::subscription_callback(
+    const sensor_msgs::msg::PointCloud2::SharedPtr lidar_cloud) {
+  pcl::PointCloud<PointXYZIRT> temp_cloud;
+  // RCLCPP_INFO(this -> get_logger(), "Header incoming: %s", lidar_cloud ->
```
Comment (Edwardius) [inline src/world_modeling/occupancy_segmentation/src/occupancy_segmentation_node.cpp:18]: instead of commenting, turn this into a DEBUG message

File: src/world_modeling/occupancy_segmentation/src/occupancy_segmentation_node.cpp [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+
+  auto begin = Clock::now();
+  _patchwork.segment_ground(temp_cloud, ground, nonground);
+  auto end = Clock::now();
+
+  int duration = std::chrono::duration_cast<std::chrono::milliseconds>(end - begin).count();
+  RCLCPP_INFO(this->get_logger(), "Runtime for Patchwork: %i ms", duration);
+  // RCLCPP_INFO(this->get_logger(), "Temp_cloud points %i", static_cast<int>(temp_cloud.size()));
+  // RCLCPP_INFO(this->get_logger(), "Ground points %i", static_cast<int>(ground.size()));
+  // RCLCPP_INFO(this->get_logger(), "Non ground points %i", static_cast<int>(nonground.size()));
```
Comment (Edwardius) [inline src/world_modeling/occupancy_segmentation/src/occupancy_segmentation_node.cpp:38]: ditto, DEBUG log RCLCPP_DEBUG

**Review by Edwardius — COMMENTED** [review]
Awesome work, gives me hope that the repo code will be very nice and clean :)

**Comment by Edwardius** [comment]
Address comments and reping me

File: src/world_modeling/occupancy_segmentation/include/occupancy_segmentation_core.hpp
Comment (akilpath) [reply src/world_modeling/occupancy_segmentation/include/occupancy_segmentation_core.hpp:1]: I did this originally, but I had linking issues when compiling and the solution I found online was to define within the hpp. I can try again and see if I can get it to work with separated definitions.

File: src/world_modeling/occupancy_segmentation/include/occupancy_segmentation_core.hpp (hunk 6 above)
Comment (akilpath) [reply src/world_modeling/occupancy_segmentation/include/occupancy_segmentation_core.hpp:72]: lmins and lmaxs won't be changed, but the others can be changed as hyperparameters for the framework. Do you want these also ROS parameterized?

File: .github/templates/docker_context/docker_context.sh (hunk 1 above)
Comment (VishGit1234) [reply .github/templates/docker_context/docker_context.sh:24]: done

File: docker/world_modeling/occupancy_segmentation/occupancy_segmentation.Dockerfile (hunk 2 above)
Comment (VishGit1234) [reply docker/world_modeling/occupancy_segmentation/occupancy_segmentation.Dockerfile]: done

File: docker/world_modeling/occupancy_segmentation/occupancy_segmentation.Dockerfile (hunk 3 above)
Comment (VishGit1234) [reply docker/world_modeling/occupancy_segmentation/occupancy_segmentation.Dockerfile]: done

File: modules/dev_overrides/docker-compose.world_modeling.yaml (hunk 4 above)
Comment (VishGit1234) [reply modules/dev_overrides/docker-compose.world_modeling.yaml:16]: done

File: src/world_modeling/occupancy_segmentation/include/occupancy_segmentation_core.hpp (hunk 5 above)
Comment (VishGit1234) [reply src/world_modeling/occupancy_segmentation/include/occupancy_segmentation_core.hpp]: done

File: src/world_modeling/occupancy_segmentation/include/occupancy_segmentation_core.hpp (hunk 6 above)
Comment (VishGit1234) [reply src/world_modeling/occupancy_segmentation/include/occupancy_segmentation_core.hpp]: done

File: src/world_modeling/occupancy_segmentation/include/occupancy_segmentation_core.hpp
Comment (VishGit1234) [reply src/world_modeling/occupancy_segmentation/include/occupancy_segmentation_core.hpp:1]: done

File: src/world_modeling/occupancy_segmentation/include/occupancy_segmentation_node.hpp (hunk 7 above)
Comment (VishGit1234) [reply src/world_modeling/occupancy_segmentation/include/occupancy_segmentation_node.hpp]: done

File: src/world_modeling/occupancy_segmentation/config/params.yaml
Comment (VishGit1234) [reply src/world_modeling/occupancy_segmentation/config/params.yaml:1]: done

File: src/world_modeling/occupancy_segmentation/src/occupancy_segmentation_node.cpp (hunk 9 above)
Comment (VishGit1234) [reply src/world_modeling/occupancy_segmentation/src/occupancy_segmentation_node.cpp]: done

File: src/world_modeling/occupancy_segmentation/package.xml (hunk 8 above)
Comment (VishGit1234) [reply src/world_modeling/occupancy_segmentation/package.xml]: done

File: src/world_modeling/occupancy_segmentation/src/occupancy_segmentation_node.cpp (hunk 10 above)
Comment (VishGit1234) [reply src/world_modeling/occupancy_segmentation/src/occupancy_segmentation_node.cpp]: done

File: src/world_modeling/occupancy_segmentation/src/occupancy_segmentation_node.cpp (hunk 11 above)
Comment (VishGit1234) [reply src/world_modeling/occupancy_segmentation/src/occupancy_segmentation_node.cpp]: done

**Review by Edwardius — APPROVED** [review]

**Comment by Edwardius** [comment]
add description of pr and we can merge!
